---
layout: cv
title: CV
permalink: /cv/
---
#### [PDF](Sridhara_BioFX.pdf)
# Viswanadham Sridhara

## CONTACT

Viswanadham.Sridhara@gmail.com  
Tel: (724) 590 2170  
[Google Scholar](https://scholar.google.com/citations?hl=en&user=UBvhRHgAAAAJ&view_op=list_works&sortby=pubdate) | [LinkedIn](https://www.linkedin.com/in/viswanadham-sridhara-308263174) | [Github](https://github.com/viswam78) | [BitBucket](https://bitbucket.org/viswam78/) | [Blog](https://sivome.github.io/)  | [DockerHub](https://hub.docker.com/u/sridharabio)



## PROFILE
[1] Experienced bioinformatics data scientist with over 15 years of experience in building bioinformatics pipelines for NGS (bulk and single-cell) and mass-spectrometry applications.
[2] Skilled in using machine learning and deep learning methods for high-dimensional biological datasets, including single-cell applications.
[3] Adept in utilizing high-performance computing methods and technologies such as snakemake, docker containerization, and cloud computing.
[4] Proven ability to work effectively in both small and large groups, utilizing agile project management methodologies.
[5] Track record of producing high-quality results with ~20 peer-reviewed publications (google scholar citations: ~875, h-index=14, incl. Nature Biomedical Engineering IF=27.0).
## EDUCATION
2003 - 2007: PHD, Electrical and Computer Engineering, ODU, Norfolk, VA
Dissertation (funded by Department of Defense MURI grant): Molecular dynamics simulations to probe effects of ultra-short high-intensity electric pulses on biological suspensions to understand how nano-second pulsed electric fields (nsPEFs) kill cancer cells.

2001 - 2003: MS, Electrical and Computer Engineering, ODU, Norfolk, VA
Thesis: Monte carlo analysis of millimeter-wave GaN Gunn diodes.

## TECHNICAL SKILLS
Programming: Python (numpy, pandas, tensorflow, keras), R (tidyverse), SQL, MATLAB
Machine learning: Regression, classification, clustering, deep learning
NGS: Bulk RNA-seq/ChIP-seq, Single cell RNA-seq/ATAC-seq, WGS (MUMmer)
Mass spectrometry: Byos, Byonic, Proteome Discoverer, MODa, ProteomeXchange
Data Engineering: Pipeline creation (Snakemake, Nextflow), feature engineering
DevOps / MLOps: Git, Docker, CI/CD, ML model deployment and optimization
Cloud Computing: AWS (e.g., EC2, S3, lambda), Google Cloud, Microsoft Azure
MD and SysBio: BioCyc, MetaCyc, SAINT, GROMACS, NAMD, VMD

## PROFESSIONAL CERTIFICATES
[1] Foundations of Project Management (Google Course)
[2] Deep learning specialization (deeplearning.ai + Coursera) - In progress
[3] Modern Big Data Analysis with SQL (Cloudera Specialization) - In progress
[4] Digital Transformation using AI/ML (Google Cloud Specialization) - In progress
## PROFESSIONAL EXPERIENCE
Jul ’22 - Feb ’23
6 months career break
• Travelled throughout India and Pacific Northwest of United States gaining cultural insights.
• Leveraged online learning platforms and resources to gain new skills and knowledge in relevant areas, including deep learning, cloud computing etc.
Apr ’21 - Jul ’22
Bioinformatics Test Engineer and Technical Writer
Protein Metrics, a Dotmatics company, Cupertino, CA

• Conducted FDR sensitivity calculations for Byonic CI/CD version testing.
• Developed and implemented automated command-line interface (CLI) tests for Byos applications, with a focus on antibody de novo assembly, hydrogen-deuterium exchange (HDX) studies, oligonucleotide identification, and other proteomics studies.
• Created new product documents, including the Byos Oligonucleotide analysis manual.
• Routinely updated product manuals (15+) and release notes for quarterly releases.
Jun ’19 - Apr '21
Senior Bioinformatics Scientist
Food and Drug Administration (Argentys contractor), Silver Spring, MD

• Built an RNA-seq pipeline for bulk processing of large-scale genomics datasets using fastp, HISAT2, featureCounts, and DESEQ2.
• Developed a single-cell RNA-seq pipeline prototype for analyzing 10X datasets using CellRanger, SAVER, Seurat/Monocle, fastMNN, and R packages.
• Characterized immune-driven drug hypersensitivity using single-cell RNA sequencing technologies.
• Identified molecular features of neural stem cells during different stages of differentiation.
• Conducted single-cell RNA-seq/ATAC-seq analyses of bone-marrow and adipose-derived MSCs using Monocle, CICERO and other related packages.
Feb ’18 - Nov '18
Bioinformatics Systems Developer
Booz Allen Hamilton, Rockville, MD

• Participated in design of cloud-based tools using EC2, S3, and Lambda (AWS technologies), to create scalable and cost-effective solutions for data storage, processing, and analysis.
• Developed a genomics pipeline using Snakemake, R packages, and Python modules for PAT, a cloud-based placental atlas tool that uses raw data from the NCBI GEO database.
• Improved analytics for differential expression, gene ontology and pathway analysis for the PAT tool (>20 RNA-seq and >300 microarray datasets).
Apr ’16 - May ’19
Independent Contractor
Universities, Hospitals and Startups (Remote)

Independent Consultant / Contractor for MGH, WRI and Mascoma
Mass General Hospital: Built an end-to-end bioinformatics pipeline and new methods to assess metrics for a long genomic region capture protocol at Mass General Hospital. Resulted in a co-first author publication in Nature Biomedical Engineering.
Windber Research Institute: Developed novel methods to integrate multi-omics data of breast cancer tissue samples at Windber Research Institute, utilizing a combination of machine learning models, clustering methods, and data from DrugBank.
Mascoma: Built custom RNA-seq pipeline, analyzed WGS data, and set-up in-house genome browser and blast+ web server for biofuel production at Mascoma.

Dec ’12 - Apr ’16
THE UNIVERSITY OF TEXAS SYSTEM (UT AUSTIN, UTSW MEDICAL CENTER DALLAS)

Research Engineering/Scientist Associate V Dec ’12 - Aug ’15
The University of Texas at Austin – CENTER FOR COMPUTATIONAL BIOLOGY AND BIOINFORMATICS

• Worked on a multi-university Army research office MURI grant to build mathematical models for multi-omics data inference.
• Managed and oversaw the development of computational infrastructure, including high-performance computing clusters, storage systems, and software tools, ensuring optimal performance and availability for research and development activities.
• Developed a method to computationally identify novel PTMs.
• Conducted 13C metabolic flux analyses for E. coli samples.
• Developed and optimized machine learning models, including LASSO (elastic net) models, to predict bacterial growth conditions based on in-silico metabolic fluxes.
• Provided proteomics bioinformatics consulting and taught computational mass-spectrometry in a 4-day course at the Big Data Summer School organized by CCBB.

Computational Biologist Aug ’15 - Apr ’16
UT Southwestern Medical Center – PROTEOMICS CORE FACILITY

• Conducted protein identification and quantitation analyses using FAIR data principles with different bioinformatics tools e.g., OMSSA, MASCOT, Sequest, ProteomeDiscoverer, MaxQuant, SAINT.
• Provided expert bioinformatics and proteomics consulting for UT Southwestern investigators (in compliance with CAP/CLIA regulations), specializing in the analysis of data related to immunology, oncology, and neurological disorders.

Jan ’12 - Dec ’12
Scientific Programmer
Samuel Roberts Noble Foundation, Ardmore, OK

• Assisted in integrating GenoLogics LIMS system with mass-spec equipment.
• Built pathway databases using MetaCyc and PathwayTools.

Jun ’07 - Jan ’12
Postdoctoral Fellow
National Center for Biotechnology Information (NCBI), Bethesda, MD

• Improved noise filtering routines and optimized data analysis algorithms for high-resolution mass spectrometry data, utilizing ML models such as linear discriminant analysis to achieve robust and accurate identification and quantification of target peptides.
• Provided user support for quantitative proteomics techniques (iTRAQ, TMT 6-plex) by adding requested enzymes, modifications and ion-types.
• Developed a new method for annotating posttranslational modifications (PTMs) on genomes using high-throughput proteomics data, protein family domain definitions, and NCBI resources. This method enables large-scale annotation of conserved PTM sites and identification of novel evolutionarily conserved PTMs.
• Developed a prototype to overcome limitations in present-generation sequence library search algorithms, such as exact precursor mass match and unknown PTMs. Presented preliminary work (oral presentation) at ASMS 2011.
Summer ’06
Student Volunteer
Old Dominion University - Aerospace Engineering, Norfolk, VA

• Contributed to the improvement of programming drives for a test bogie run on a magnetic levitation train (MAGLEV).

Summer ’04
Intern
Siemens Limited. Siemens Vdo Automotive, Chesapeake, VA

• Conducted finite element analyses (2D) using Ansoft software to improve fuel injection solenoid models.

Aug ’03 - May ’07
OLD DOMINION UNIVERSITY (JOSHI LAB, FRANK REIDY RESEARCH CENTER FOR BIOELECTRICS)

Graduate Research Assistant Aug ’03 - May ’07
COMPUTATIONAL RESEARCH LAB - BIOELECTRICS, Norfolk, VA
FRANK REIDY RESEARCH CENTER FOR BIOELECTRICS, Norfolk, VA

• Conducted research under DoD MURI grant (multi university, incl. MIT) to model and simulate the effect of nano-second pulsed electric fields (nsPEFs) on cancer cells.
• Conducted intense Molecular Dynamics simulations to understand nanopore formation in lipid bilayers and computed permittivities and diffusion coefficients of local lipid membranes.
• Studied the effects of sodium and calcium ions on pore formation in biological suspensions using MD simulations.
• Mentored and trained graduate students, providing guidance and support for research projects and career development.
• Conducted Time Domain Dielectric Transient Spectroscopy measurements to monitor structural changes in leukemic cells after nsPEF application.
SELECTED TALKS
Engineered DNA Capture Probes for Massively Multiplexed Cloning of Kilobase-Sized Genome Regions, presented at FDA.

Using subspectral interval matches to make novel identifications of ETD tandem mass spectra, presented at the 2011 American Society for Mass Spectrometry (ASMS) conference.

Predicting bacterial growth conditions from bacterial physiology, presented at the Center for Systems and Synthetic Biology at UT-Austin.

## PUBLICATIONS
Tosi, L., Sridhara, V., et. al., Nature Biomedical Engineering, 2017.

Caglar, M. U., et. al., Scientific Reports 2017.

Brown, C. W., Sridhara, V., et. al., BMC Genomics 2017.

Kalantari, R., et. al., RNA 2016.

Houser, J. R., et. al., PLOS Computational Biology 2015.

Sridhara, V., et. al., PLoS One 2014.

Sridhara, V., Joshi, R. P. Biochimica Biophysica Acta - Biomembranes 2014.

Sridhara, V., Joshi, R. P. Biochimica Biophysica Acta - Biomembranes 2014.

Sridhara, V., et. al., Proteome Science 2012.

Sridhara, V., et. al., Database (Oxford) 2011.

Joshi, R.P., et. al., IEEE Trans. Dielectrics & Electrical Insulation 2009.

Allen L. Garner, et. al., Biochemical Biophysical Research Communications 2007.

Joshi, R.P., et. al., Physical review. E, Statistical, nonlinear, and soft matter physics 2007.

Pliquett, U., et. al., Bioelectrochemistry 2006.

Joshi, R.P., Sridhara, V., et. al., Biochemical Biophysical Research Communications 2006.

Qin Hu, Sridhara, V., et. al., (invited paper) IEEE Transactions on Plasma Science 2006.

Hu, Q., Sridhara, V., et. al., Physical Review E - Statistical, Nonlinear, and Soft Matter Physics 2005.

Joshi, R. P., Sridhara, V., et. al., Journal of Applied Physics 2003.

Joshi, R. P., Sridhara, V., et. al., Journal of Applied Physics 2003.

## CONFERENCE PROCEEDINGS
Sridhara, V., et. al., (invited paper) Bio-Electromagnetic Society Conference 2006.

Pliquett, U., et. al., Electromed 2005 Conference.

Garner, A.L., et. al., Electromed 2005 Conference.

## POSTERS
SCD(FDA) 2020, AACR 2017, Proceedings of the AACR 2016, EMBL Conference 2014, Synthetic Biology: Engineering, Evolution and Design (SEED), ASMS '08-'10

## GRANT REVIEWER
National Science Foundation (NSF)

## JOURNALS REFEREED
JPR, PLoS Pathogens, Journal of Theoretical Biology, Biology Direct, PLoS Computational Biology, International Journal of Bioinformatics Research and Applications

## PROFESSIONAL MEMBERSHIPS
American Association for Cancer Research, American society for mass spectrometry

## AWARDS
• 2007: NIH Postdoctoral Fellowship
• 2007: Outstanding Graduate Researcher Award, Old Dominion University
• 2006: BCET Scholarship, Old Dominion University
• 2006: Honorary Mention, ODU Research Foundation Day Poster Presentation
• 1996-2000: National Merit Scholarship, College of Engineering, Osmania University
• 1994-1996: College Scholarship, St. Thomas Samhita Jr. College
• 1994: Certificate of Merit in Mathematics

## SAMPLE PROJECTS
Machine Learning
PLoS One paper - Regularization and metabolic network models.
Python Jupyter notebook (FBA and sklearn) | R markdown GLMNET analysis

Genomics
Nature Biomedical Engineering paper - Designing probes for genome capture.
Reads to BED format to metrics to plots analysis using R markdown

Cancer Multi-omics
AACR sympsoium [URL: https://sivome.github.io/Poster_PDFs/Proteogenomic_AACR_Poster.pdf]

Single-cell work
SCD (FDA) poster [URL: https://sivome.github.io/Poster_PDFs/SingleCellMSC_SCD_Poster.pdf]

Deep Learning
Neural Networks - 10X single cell data using Python Jupyter notebook [URL: https://sivome.github.io/genomics/2019/11/17/Denoising-Single-Cell-Data-With-Autoencoders.html]
