---
layout: page
title: CV
permalink: /cv/
---
# [BioFX-PDF](BioFX_Sridhara.pdf)  
# [DS-PDF](DS_Sridhara.pdf)  
# Viswanadham (Vish) Sridhara

## CONTACT

Viswanadham.Sridhara@gmail.com  
[Google Scholar](https://scholar.google.com/citations?hl=en&user=UBvhRHgAAAAJ&view_op=list_works&sortby=pubdate) | [LinkedIn](https://www.linkedin.com/in/viswanadham-sridhara-308263174) | [Github](https://github.com/viswam78) | [BitBucket](https://bitbucket.org/viswam78/) | [Blog](https://sivome.github.io/)  



## PROFILE

I am a computational scientist with modeling and simulation experience in engineering and life sciences domain. Post-PHD, I have 12+ years’ experience mining and analyzing high-dimensional structured and unstructured datasets using sophisticated machine learning models. Till date, my research work has resulted in ~20 peer-reviewed publications and numerous conference proceedings (google scholar citations: ~580, h-index=11).



## EDUCATION

2003 - 2007, PhD, Electrical and Computer Engineering, ODU, Norfolk, VA  
**Dissertation:** Molecular dynamics simulations to probe effects of ultra-short high-intensity electric pulses on biological suspensions  

2001 - 2003, MS, Electrical and Computer Engineering, ODU, Norfolk, VA  
**Thesis:** Monte carlo analysis of millimeter-wave GaN Gunn diodes  


## TECHNICAL SKILLS

Machine learning: Regression, Regularization, Hierarchical Clustering, Discriminant analyses  
Programming Languages: Python (numpy, pandas, scikit-learn, matplotlib, keras), R (tidyverse), MATLAB, Perl, C, C++, Fortran   
Other skills: Git, Snakemake, Docker, UNIX, MacOS, LaTeX  
Engineering focussed : Monte-carlo models [MS Thesis focus], Molecular dynamics simulations for atomic modeling [PHD Disseration focus]

## SAMPLE PROJECTS  
[Technical sample blog with Python Jupyter notebook](https://sivome.github.io/proteomics/2019/03/11/Reporting_Proteomics_Results.html)  
[Technical sample blog with R markdown](https://sivome.github.io/proteomics/2019/03/15/universe-of-proteomics-with-tidyverse.html)  
[Paper - Linear Discriminant Analyses](https://www.ncbi.nlm.nih.gov/pubmed/22321509)  
[Paper - LASSO regularization](https://www.ncbi.nlm.nih.gov/pubmed/25502413)  
[Paper - Visualization using R tidyverse](https://www.ncbi.nlm.nih.gov/pubmed/29152409)  


## PROFESSIONAL EXPERIENCE  


### Consultant
**Universities, Hospitals and Startups**  
Oct 2018 - Present  (Sivome LLC)
In my current role, I am working for a proteomics software company.

In previous roles as Independent Consultant (part of 2016, 2017) for MGH, WRI and biofuel startup  
• Improved primer design for a proof-of-principle assay to capture long genomic regions.  
• Built NGS pipeline and developed new computational methods for downstream analyses.  
• Set-up in-house Jbrowse genome browser and Sequenceserver (blast+ web server).  
• Helped analyze diverse mass-spectrometry datasets (global proteomics, phosphoproteomics) of breast cancer tissue samples.  


### Bioinformatics Systems Developer      
**Booz Allen Hamilton**  
March 2018 - November 2018    

• Helped in automating preprocessing of gene expression data (microarray and RNA-seq) for placental datasets using raw data stored in NCBI’s GEO repository (PAT - Placental Atlas Tool).  
• Improved analytics of differential expression, gene ontology and pathway analyses of the above preprocessed placental datasets (>20 RNA-seq and >300 microarray datasets).


### Computational Biologist
**UT Southwestern Medical Center -- PROTEOMICS CORE FACILITY**  
August 2015 - April 2016  

• Protein identification and quantitation using bioinformatics tools.  
• Bioinformatics proteomics consultant work with UT Southwestern investigators.  

### Research Engineering/Scientist Associate V
**The University of Texas at Austin -- CENTER FOR COMPUTATIONAL BIOLOGY AND BIOINFORMATICS**  
December 2012 - August 2015  

 • Built machine learning models using LASSO regression to predict bacterial growth conditions from _in-silico_ metabolic fluxes generated from metabolic network based flux balance analyses (FBA) methods.  
 • Proteomics and metabolic flux data analyses of mass-spectrometry based datasets.
   &nbsp;&nbsp;a. Proteomics data is analyzed using new unrestricted PTM search methods.
   &nbsp;&nbsp;b. Metabolic flux data is analyzed using FiatFlux software.  
 • Molecular dynamics (MD) simulations for Bio-electrics applications  
  &nbsp;&nbsp;a. Compute intense MD simulations to understand multiple nanopore formation in lipid bilayers using advanced high performance computing environment.  
  &nbsp;&nbsp;b. MD simulations to understand effects of sodium (Na) and calcium (Ca) effects in pore formation within the biological suspensions.  
 • Helped multiple labs with proteomics bioinformatics consulting and short-term contracts.  
 • Taught computational mass-spectrometry in Big Data Summer School organized by CCBB (Center for Computational Biology and Bioinformatics every summer – 4 day course)  

### Visiting Research Fellow
**National Center for Biotechnology Information (NCBI)**  
June 2007 - January 2012  

 • Part of NCBI OMSSA (Open Mass Spectrometry Search Algorithm) project.  
 • Integration of sequence data & high-throughput phosphoproteomics data.  
 • Developed a prototype for new class of peptide identification search algorithm OMSSA-SIM.  


### Student Volunteer  
**Old Dominion University - Aerospace Engineering**  
Summer '06
• Helped improve programming drives for a test bogie run on MAGLEV (train using magnetic levitation).


### Intern
**Siemens Limited. Siemens Vdo Automotive**  
Summer '04
• Finite element analyses (2D) to improve fuel injection solenoid models using Ansoft software.  


## SELECTED TALKS

1. Large-scale analysis of PTMs @ UT-Austin, Thermo Fisher Scientific, MD Anderson Cancer Center

2. Using subspectral interval matches to make novel identifications of ETD tandem mass spectra @ ASMS 2011

3. Predicting bacterial growth conditions from bacterial physiology @ Center for Systems and Synthetic Biology, UT-Austin

4. Open Mass Spectrometry Search Algorithm @ George Washington University  

## PUBLICATIONS

1. Tosi, L., Sridhara, V., Yang, Y., et. al., “Engineered DNA Capture Probes for Massively Multiplexed Cloning of Kilobase-Sized Genome Regions”, Nature Biomedical Engineering, 2017;1. pii: 0092. doi: 10.1038/s41551-017-0092. Epub 2017 Jun 26.

2. Caglar, M. U., et. al.,  “The E. coli molecular phenotype under different growth conditions”, Scientific Reports 2017 Apr 18;7:45303. doi: 10.1038/srep45303.

3. Brown, C. W., Sridhara, V., Boutz, D. R., et. al., “Large-scale analysis of post-translational modifications in E. coli under glucoselimiting conditions”, BMC Genomics  2017 Apr 17;18(1):301. doi: 10.1186/s12864-017-3676-8.

4. Kalantari, R., Hicks, J., Li, L., Gagnon, K., Sridhara, V., Lemoff, A., Mirzaei, H., and Corey, D. R. “Stable Association of RNAi Machinery is Conserved Between the Cytoplasm and Nucleus of Human Cells”, RNA. 2016 Jul;22(7):1085-98. doi: 10.1261/rna.056499.

5. Houser, J. R., et. al., Controlled measurement and comparative analysis of cellular components in E. coli reveals broad regulatory changes under long-term starvation. PLOS Computational Biology 2015 Aug 14;11(8):e1004400.

6. Sridhara, V., Meyer AG, Rai P, Barrick JE, Ravikumar P, Segre D, and Wilke C.O. “ Predicting growth conditions from internal metabolic fluxes in an in-silico model of E. coli.” (PLoS One.
2014 Dec 12;9(12):e114608.).

7. Sridhara, V., Joshi, R. P. “Evaluations of a mechanistic hypothesis for the influence of extracellular ions on electroporation due to high-intensity, nanosecond pulsing.” Biochimica Biophysica Acta - Biomembranes, Vol. 1838, no. 7, Jul 2014, pp. 1793-1800.  

8. Sridhara, V., Joshi, R. P. “Numerical study of lipid translocation driven by nanoporation due to multiple high-intensity, ultrashort electrical pulses.” Biochimica Biophysica Acta - Biomembranes, Vol. 1838, no. 7, Mar 2014, pp. 902-909.

9. Sridhara, V., Dina L. Bai, An Chi, Jeffrey Shabanowitz, Donald F. Hunt, Stephen H. Bryant and Lewis Y. Geer “Increasing peptide identifications and decreasing search times for ETD spectra by pre-processing and calculation of parent precursor charge.”  Proteome Science 2012, 10:8.  

10. Sridhara, V., Aron Marchler-Bauer, Stephen H. Bryant and Lewis Y. Geer, “Automatic annotation of experimentally derived, evolutionarily conserved post-translational modifications onto multiple genomes.” Database (Oxford), 2011, bar019.  

11. Joshi, R.P., Sridhara, V., Song, J., Schoenbach, K.H  “Aspects of Lipid Membrane BioResponses to Subnanosecond, Ultrahigh Voltage Pulsing.” IEEE Trans. Dielectrics & Electrical Insulation, Vol. 16, no. 5, Oct 2009, pp. 1243-1250.  

12. Allen L. Garner, George Chen, Nianyong Chen, Viswanadham Sridhara, Juergen F. Kolb, R. James Swanson, Stephen J. Beebe, Ravindra P. Joshi, and Karl H. Schoenbach “Changes in the Dielectric Properties of Jurkat Cells Induced by Ultrashort Electrical Pulses,” Biochemical Biophysical Research Communications Vol. 362, Issue. 1, Oct. 2007, pp.  139-144.  

13. Joshi, R.P., Nguyen, A., Sridhara, V., Nuccitelli, R., and  Schoenbach, K.H., “Simulations of Calcium Release Dynamics in Response to a High-Intensity,  Ultra-Short Electric Pulse,” Physical review. E, Statistical, nonlinear, and soft matter physics 2007; 75(4 Pt 1):041920.  

14. Pliquett, U., Joshi, R.P., Sridhara, V., Schoenbach, K.H. “High Electrical Field effects on Cell Membranes.” Bioelectrochemistry Vol. 70, Oct. 2006, pp. 275-282.  

15. Joshi, R.P., Sridhara, V., Schoenbach, K.H  “Microscopic Calculations of Local Lipid Membrane Permittivities and Diffusion Coefficients for Application to Electroporation Analyses.” Biochemical Biophysical Research Communications Vol. 348, Issue. 2, Sept. 2006, pp.  643648.  

16. Qin Hu, Sridhara, V., R. P. Joshi, J. Kolb and K.H, Schoenbach, “Molecular Dynamics Analysis of High Electric Pulse effects on Bilayer Membranes Containing DPPC and DPPS.”  (invited paper) IEEE Transactions on Plasma Science, Vol. 34, no. 4, Aug. 2006, pp.  1405-1411.  

17. Hu, Q., Sridhara, V., Joshi, R.P., Schoenbach, K.H., Beebe, S.J., Blackmore, P.F. “Simulations of transient membrane behavior in cells subjected to a high-intensity ultrashort electric pulse.” Physical Review E - Statistical, Nonlinear, and Soft Matter Physics, Vol. 71,no. 3, March 2005, pp. 031914(1-7).  

18. Joshi, R. P., Sridhara, V., Jogai, B., Shah, P., Del Rosario, R.D. “Analysis of dislocation scattering on electron mobility in GaN high electron mobility transistors.” Journal of Applied Physics, Vol. 93, no. 12, Jun 15, 2003, pp. 10046-10052.  

19. Joshi, R. P., Sridhara, V., Shah, P., Del Rosario, R.D. “Monte Carlo analysis of GaN-based Gunn oscillators for microwave power generation.” Journal of Applied Physics, Vol. 93, no. 8, Apr 15, 2003, pp. 4836-4842.


## CONFERENCE PROCEEDINGS

1. Sridhara, V., Joshi, R.P., Hu, Q., Nguyen, A., and Schoenbach, K.H. “Simulations of cell response to high intensity, ultrashort electrical pulses” (invited paper) Bio-Electromagnetic Society Conference, Cancun, Mexico, June 11-15, 2006.

2. Pliquett, U., et. al., “Monitoring intrastructural changes in Jurkat cells after nanosecond pulsed electric field application.” Electromed 2005 Conference, Portland, OR, May 15-18 , 2005.

3. Garner, A.L., et. al., “Changes in the Dielectric Properties of Jurkat and HL-60 cells after Electrical Pulses.” Electromed 2005 Conference, Portland, OR, May 15-18, 2005.

## POSTERS

AACR 2017, Proceedings of the AACR 2016, EMBL Conference 2014, Synthetic Biology: Engineering, Evolution and Design (SEED), ASMS 2010, ASMS 2009, ASMS 2008
