<!doctype html>
<html>
<head>
  <meta charset=utf-8 />
  <link href="media/vish-screen.css" type="text/css" rel="stylesheet" media="screen">
  <link href="media/vish-print.css" type="text/css" rel="stylesheet" media="print">
</head>
<body>
  <div id="main">
    <div id="content">
<h1 id="viswanadham-sridhara">Viswanadham (Vish) Sridhara</h1>
<h2 id="contact">CONTACT</h2>
<p>Viswanadham.Sridhara@gmail.com<br />
Tel: (724) 590 2170<br />
<p>Current role: Senior Bioinformatics Scientist<br />
<p>Citizenship: USA<br />
<a href="https://scholar.google.com/citations?hl=en&amp;user=UBvhRHgAAAAJ&amp;view_op=list_works&amp;sortby=pubdate">Google Scholar</a> | <a href="https://www.linkedin.com/in/viswanadham-sridhara-308263174">LinkedIn</a> | <a href="https://github.com/viswam78">Github</a> | <a href="https://bitbucket.org/viswam78/">BitBucket</a> | <a href="https://sivome.github.io/">Blog</a> | <a href="https://hub.docker.com/u/sridharabio">DockerHub</a></p>
<h2 id="profile">PROFILE</h2>
<p>Project management skills with experience in developing bioinformatics methods and building bioinformatics pipelines for NGS and mass-spec technologies. Recent focus is on single-cell analyses and ML methods such as deep learning. </p><br />
<p>Comfortable using high-performance (and cloud) computing methods in *nix environment coupled with workflow manager. Proficient in Python/R/MATLAB.</p><br />
<p>Worked for industry, government and academia. Past research work is funded by Army research office (ARO MURI grant), Department of Defense (DoD MURI grant) and National Institutes of Health that led to additional grants for PIs.
</p>
<h2 id="education">EDUCATION</h2>
<h3 id="section">2003 - 2007</h3>
<p>PhD, Electrical and Computer Engineering, ODU, Norfolk, VA<br />
<strong>Dissertation (funded by Department of Defense MURI grant):</strong> Molecular dynamics simulations to probe effects of ultra-short high-intensity electric pulses on biological suspensions to understand how nano-second pulsed electric fields (nsPEFs) kill cancer cells.</p>
<h3 id="section-1">2001 - 2003</h3>
<p>MS, Electrical and Computer Engineering, ODU, Norfolk, VA<br />
<strong>Thesis:</strong> Monte carlo analysis of millimeter-wave GaN Gunn diodes</p>
<h2 id="technical-skills">TECHNICAL SKILLS</h2>
<h3 id="machine-learning">Machine learning</h3>
<p>Regression, Regularization, Hierarchical Clustering, Discriminant analyses<br />
<h3 id="programming">Programming</h3>
<p>Python (numpy, pandas, scikit-learn, matplotlib, keras), R (tidyverse, ML packages), MATLAB, Perl, UNIX, MacOS, LaTeX<br />
<h3 id="other-skills">BioFX Pipeline</h3>
<p>Cloud computing, Git, Snakemake, Docker<br />
<h3 id="engineering">Engineering</h3>
<p>Monte-carlo models, Molecular dynamics simulations 
<h3 id="ngs">NGS</h3>
<p>Bulk RNA-seq: HISAT2, featureCounts, GATK, BWA, Bowtie, Samtools, BEDtools (familiarity with TCGC, ICGC datasets)<br />
<p>Single cell RNA-seq: CellRanger, Seurat, Monocle3, fastMNN, SAVER<br /> 
<h3 id="mass-spec">Mass-spec</h3> 
<p>Byonic, PEAKS, Proteome Discoverer, OMSSA, MODa (familiarity with CPTAC, Tranche, PRIDE, PeptideAtlas datasets)<br />
<h3 id="systems-biology">Systems Biology</h3>
<p>BioCyc, MetaCyc, SAINT 
<h3 id="md-simulations">MD Simulations</h3>
<p>GROMACS, NAMD, VMD</p>
<h2 id="professional-experience">PROFESSIONAL EXPERIENCE</h2>
<h3 id="jun-19---present">Jun ’19 - Present</h3>
<p><em>Senior Bioinformatics Scientist</em><br />
<strong>Argentys (FDA, universities and hospitals contract work) </strong></p>
<p>Contractor at Food and Drug Administration (Jun’ 19 - Aug’ 20) CBER HIVE Team</p>
<p><strong>Informatics Projects at FDA</strong></p>
<p>• Helped in building RNA-seq pipeline for automated bulk processing of large-scale genomics datasets (fastp, HISAT2, featureCounts, DESEQ2)<br />
• Built single-cell RNA-seq workflow for analyzing 10X single cell datasets (CellRanger, SAVER, Seurat/Monocle, fastMNN and other R packages)</p>
<p><strong>Sample Research Projects at FDA</strong></p>
<p>• Characterization of the Immune-Driven Drug Hypersensitivity by Single-Cell RNA Sequencing Technologies<br />
• Identifying Molecular Features of Neural Stem Cells in Varied Differentiation Stages<br />
• Single cell RNA-seq analyses to investigate sub-populations of bone-marrow and adipose derived multipotential stromal cells (MSCs) from 3 donors</p>
<br>
<p> <strong>Universities and medical centers (current contract)</strong></p>
<p>• NGS analyses of targeted LINE-1 retrotransposons data <br />
• NGS analyses to understand vaccine stability <br />  
(took a 5 month break to stay in India between FDA role and current contract) </p>

<h3 id="apr-16---may-19">Apr ’16 - May ’19</h3>
<p><em>Independent Contractor</em><br />
<strong>Universities, Hospitals and Startups</strong></p>
<p>Independent Consultant / Contractor for MGH, WRI, biofuel startup, Booz Allen Hamilton and ProteinMetrics<br />
• At ProteinMetrics, tested ProteinMetrics suite of tools.<br /> 
• At Booz Allen Hamilton, worked on Placental atlas tool (PAT), a meta-data driven web interface to analyze gene expression data on the fly. Automated workflow and improved data analytics to analyze gene expression data (300 microarray datasets and 20 RNA-seq datasets - 1000's of samples)<br />
• At Mass General Hospital, developed bionformatics methods for an assay to capture and clone long dna fragments in high-throughput and cost effective way. [Publised in Nature Biomedical Engineering]<br />
• At biofuel startup, whole genome sequence (WGS) analyses, including variant analyses, copy number variation and strain specific analyses for biofuel production. Additional work to set-up in-house Jbrowse genome browser and sequence server<br />
• At WRI (in collaboration with PNNL), integrated proteogenomics data (>100 RNA-seq and >50 mass-spec) to identify markers in laser capture microdissected breast tumor samples (Preliminary work published in AACR).</p>
<h3 id="dec-12---apr-16">Dec ’12 - Apr ’16</h3>
<p><strong>THE UNIVERSITY OF TEXAS SYSTEM (UT AUSTIN, UTSW MEDICAL CENTER DALLAS)</strong></p>
<p><em>Research Engineering/Scientist Associate V</em> Dec ’12 - Aug ’15<br />
<strong>The University of Texas at Austin – CENTER FOR COMPUTATIONAL BIOLOGY AND BIOINFORMATICS</strong></p>
<p>
• Army research office MURI grant (multi university, incl. Boston University) to build mathematical models to infer associations in multi-omics data <br />
  a. Novel PTM identification using Unrestricted bioinformatic search of large-scale mass-spectrometry based proteomics data generated for mutiple E. coli growth conditions. <br />
  b. 13C Metabolic flux analyses for the same above mentioned samples using FiatFlux and other software.<br />
• Built machine learning models using LASSO regression to predict bacterial growth conditions from <em>in-silico</em> metabolic fluxes generated from metabolic network based flux balance analyses (FBA) methods.<br />
• Helped multiple labs with proteomics bioinformatics consulting and short-term contracts.<br />
• Taught computational mass-spectrometry in Big Data Summer School organized by CCBB (Center for Computational Biology and Bioinformatics every summer – 4 day course)</p>
<br>
<p><em>Computational Biologist</em> Aug ’15 - Apr ’16<br />
<strong>UT Southwestern Medical Center – PROTEOMICS CORE FACILITY</strong></p>
<p>• Protein identification and quantitation using bioinformatics tools (e.g., OMSSA, MASCOT, Sequest, ProteomeDiscoverer, MaxQuant, SAINT).<br />
• Bioinformatics proteomics consultant work with UT Southwestern investigators.</p>

<h3 id="jan-12---dec-12">Jan ’12 - Dec ’12</h3>
<p><em>Scientific Programmer</em><br />
<strong>Samuel Roberts Noble Foundation</strong></p>
<p>• Helped integrated GenoLogics LIMS system with mass-spec equipment • Built pathway databases using MetaCyc and PathwayTools</p>
<h3 id="jun-07---jan-12">Jun ’07 - Jan ’12</h3>
<p><em>Postdoctoral Fellow</em><br />
<strong>National Institutes of Health (NCBI/NIH) - Computational Mass Spectrometry </strong></p>
<p>• OMSSA User support by addition of requested enzymes, modifications, and other iontypes required for quantitative proteomics (iTRAQ, TMT 6-plex etc.). ETD / OMSSA work presented in ASMS ’08, ’09, ’10 and ’11.</p>
<p>
<p>• Developed a new method (proof-of-principle) for PTM annotation on genomes (NCBI RefSeq) by using high-throughput proteomics data (PhosphoSite, PhosphoELM) and domain definitions of protein families within NCBI Conserved Domain Database</p>
<p>
<p>• Developed a prototype for new class of peptide identification search algorithm OMSSA-SIM.</p>
<h3 id="summer-06">Summer ’06</h3>
<p><em>Student Volunteer</em><br />
<strong>Old Dominion University - Aerospace Engineering</strong></p>
<p>• Helped improve programming drives for a test bogie run on MAGLEV (train using magnetic levitation).</p>
<h3 id="summer-04">Summer ’04</h3>
<p><em>Intern</em><br />
<strong>Siemens Limited. Siemens Vdo Automotive</strong></p>
<p>• Finite element analyses (2D) to improve fuel injection solenoid models using Ansoft software.</p>
<h3 id="aug-03---may-07">Aug ’03 - May ’07</h3>
<p><strong>OLD DOMINION UNIVERSITY / FRANK REIDY RESEARCH CENTER FOR BIOELECTRICS</strong></p>
<p><em>Graduate Research Assistant</em> Aug ’03 - May ’07<br />
<strong>ODU COMPUTATIONAL RESEARCH LAB - BIOELECTRICS</strong></p>
<p>• DoD MURI grant (multi university, incl. MIT) to model/simulate how nano-second pulsed electric fields (nsPEFs) kill cancer cells<br />
a. Compute intense Molecular Dynamics (MD) simulations to understand multiple nanopore formation in lipid bilayers using advanced high performance computing environment.<br />
b. MD simultations for microscopic calculation of local lipid membrane permittivities and diffusion coefficients, and to understand effects of sodium (Na) and calcium (Ca) ions in pore formation within the biological suspensions.</p>
<p>
<br>
<p><em>Research Assistant V</em> May ’04 - Dec ’06<br />
<strong>FRANK REIDY RESEARCH CENTER FOR BIOELECTRICS</strong></p>
<p>• Time Domain Dielectric Transient Spectroscopy (TDDS) measurements to monitor intrastructural changes in leukaemic cells after nanosecond pulsed electric field application.</p>
<h2 id="grant reviewer">GRANT REVIEWER</h2>
<p>National Science Foundation (NSF)</p>
<h2 id="journals refereed">JOURNALS REFEREED</h2>
<p>Journal of Proteome Research, PLoS Pathogens, Journal of Theoretical Biology, Biology Direct, PLoS Computational Biology</p>
<h2 id="professional memberships">PROFESSIONAL MEMBERSHIPS</h2>
<p>• American Association for Cancer Research <br />
• American society for mass spectrometry
</p>
<h2 id="awards">AWARDS</h2>
<p>• June 2007: NIH Postdoctoral Fellowship <br />
• April 2007: Outstanding Graduate Researcher Award (Ph. D.) <br />
• June 2006: Old Dominion University BCET Scholarship <br />
• April 2006: Honorary Mention for ODU Research Foundation Day Poster Presentation <br />
• 1996-2000: National Merit Scholarship, College of Engineering, Osmania University <br />
• 1994-1996: College Scholarship, St. Thomas Samhita Jr. College <br />
• May 1994: Certificate of merit in Math
</p>  
<h2 id="selected-talks">SELECTED TALKS</h2>
<ol type="1">
<li><p>Large-scale analysis of PTMs @ UT-Austin, Thermo Fisher Scientific, MD Anderson Cancer Center</p></li>
<li><p>Using subspectral interval matches to make novel identifications of ETD tandem mass spectra @ ASMS 2011</p></li>
<li><p>Predicting bacterial growth conditions from bacterial physiology @ Center for Systems and Synthetic Biology, UT-Austin</p></li>
<li><p>Open Mass Spectrometry Search Algorithm @ George Washington University</p></li>
</ol>
<h2 id="publications">PUBLICATIONS</h2>
<ol type="1">
<li><p>Tosi, L., Sridhara, V., et. al., “Engineered DNA Capture Probes for Massively Multiplexed Cloning of Kilobase-Sized Genome Regions”, Nature Biomedical Engineering, 2017.</p></li>
<li><p>Caglar, M. U., et. al., “The E. coli molecular phenotype under different growth conditions”, Scientific Reports 2017.</p></li>
<li><p>Brown, C. W., Sridhara, et. al., “Large-scale analysis of post-translational modifications in E. coli under glucoselimiting conditions”, BMC Genomics 2017.</p></li>
<li><p>Kalantari, R., Hicks, et. al.,  “Stable Association of RNAi Machinery is Conserved Between the Cytoplasm and Nucleus of Human Cells”, RNA. 2016.</p></li>
<li><p>Houser, J. R., et. al., Controlled measurement and comparative analysis of cellular components in E. coli reveals broad regulatory changes under long-term starvation. PLOS Computational Biology 2015.</p></li>
<li><p>Sridhara, V., Meyer AG, Rai P, Barrick JE, Ravikumar P, Segre D, and Wilke C.O. “ Predicting growth conditions from internal metabolic fluxes in an in-silico model of E. coli.” PLoS One 2014.</p></li>
<li><p>Sridhara, V., Joshi, R. P. “Evaluations of a mechanistic hypothesis for the influence of extracellular ions on electroporation due to high-intensity, nanosecond pulsing.” Biochimica Biophysica Acta - Biomembranes, 2014.</p></li>
<li><p>Sridhara, V., Joshi, R. P. “Numerical study of lipid translocation driven by nanoporation due to multiple high-intensity, ultrashort electrical pulses.” Biochimica Biophysica Acta - Biomembranes, 2014.</p></li>
<li><p>Sridhara, V., Dina L. Bai, An Chi, Jeffrey Shabanowitz, Donald F. Hunt, Stephen H. Bryant and Lewis Y. Geer “Increasing peptide identifications and decreasing search times for ETD spectra by pre-processing and calculation of parent precursor charge.” Proteome Science 2012.</p></li>
<li><p>Sridhara, V., et. al., “Automatic annotation of experimentally derived, evolutionarily conserved post-translational modifications onto multiple genomes.” Database (Oxford), 2011.</p></li>
<li><p>Joshi, R.P., Sridhara, V., et. al., “Aspects of Lipid Membrane BioResponses to Subnanosecond, Ultrahigh Voltage Pulsing.” IEEE Trans. Dielectrics &amp; Electrical Insulation, 2009.</p></li>
<li><p>Allen L. Garner, et. al., “Changes in the Dielectric Properties of Jurkat Cells Induced by Ultrashort Electrical Pulses,” Biochemical Biophysical Research Communications 2007.</p></li>
<li><p>Joshi, R.P., et. al., “Simulations of Calcium Release Dynamics in Response to a High-Intensity, Ultra-Short Electric Pulse,” Physical review. E, Statistical, nonlinear, and soft matter physics 2007.</p></li>
<li><p>Pliquett, U., et. al., “High Electrical Field effects on Cell Membranes.” Bioelectrochemistry 2006.</p></li>
<li><p>Joshi, R.P., Sridhara, V., Schoenbach, K.H “Microscopic Calculations of Local Lipid Membrane Permittivities and Diffusion Coefficients for Application to Electroporation Analyses.” Biochemical Biophysical Research Communications 2006.</p></li>
<li><p>Qin Hu, Sridhara, V., et. al.,, “Molecular Dynamics Analysis of High Electric Pulse effects on Bilayer Membranes Containing DPPC and DPPS.” (invited paper) IEEE Transactions on Plasma Science 2006.</p></li>
<li><p>Hu, Q., Sridhara, V., et. al., “Simulations of transient membrane behavior in cells subjected to a high-intensity ultrashort electric pulse.” Physical Review E - Statistical, Nonlinear, and Soft Matter Physics, 2005.</p></li>
<li><p>Joshi, R. P., Sridhara, V., Jogai, B., Shah, P., Del Rosario, R.D. “Analysis of dislocation scattering on electron mobility in GaN high electron mobility transistors.” Journal of Applied Physics, 2003.</p></li>
<li><p>Joshi, R. P., Sridhara, V., Shah, P., Del Rosario, R.D. “Monte Carlo analysis of GaN-based Gunn oscillators for microwave power generation.” Journal of Applied Physics, 2003.</p></li>
</ol>
<h2 id="conference-proceedings">CONFERENCE PROCEEDINGS</h2>
<ol type="1">
<li><p>Sridhara, V., Joshi, R.P., Hu, Q., Nguyen, A., and Schoenbach, K.H. “Simulations of cell response to high intensity, ultrashort electrical pulses” (invited paper) Bio-Electromagnetic Society Conference, Cancun, Mexico, June 11-15, 2006.</p></li>
<li><p>Pliquett, U., et. al., “Monitoring intrastructural changes in Jurkat cells after nanosecond pulsed electric field application.” Electromed 2005 Conference, Portland, OR, May 15-18 , 2005.</p></li>
<li><p>Garner, A.L., et. al., “Changes in the Dielectric Properties of Jurkat and HL-60 cells after Electrical Pulses.” Electromed 2005 Conference, Portland, OR, May 15-18, 2005.</p></li>
</ol>
<h2 id="posters">POSTERS</h2>
<p>SCD(FDA) 2020, AACR 2017, Proceedings of the AACR 2016, EMBL Conference 2014, Synthetic Biology: Engineering, Evolution and Design (SEED), ASMS 2010, ASMS 2009, ASMS 2008</p>
<h2 id="sample-projects">SAMPLE PROJECTS</h2>
<h3 id="ml-lda">ML-LDA</h3>
<p><a href="https://www.ncbi.nlm.nih.gov/pubmed/22321509">Database (Oxford) paper summarizing ML + bioinformatics [URL: https://www.ncbi.nlm.nih.gov/pubmed/22321509]</a>
<h3 id="ml-regularization">ML-Regularization</h3>
<p><a href="https://www.ncbi.nlm.nih.gov/pubmed/25502413">PLoS One paper - Regularization and metabolic network models [URL: https://www.ncbi.nlm.nih.gov/pubmed/25502413]</a>
<h3 id="cancer-proteogenomics">Cancer-Proteogenomics</h3> 
<p><a href="https://sivome.github.io/Poster_PDFs/Proteogenomic_AACR_Poster.pdf">AACR poster [URL: https://sivome.github.io/Poster_PDFs/Proteogenomic_AACR_Poster.pdf]</a></p>
<h3 id="single-cell-fda-project">single-cell FDA-project</h3> 
<p><a href="https://sivome.github.io/Poster_PDFs/SingleCellMSC_SCD_Poster.pdf">SCD (FDA) poster [URL: https://sivome.github.io/Poster_PDFs/SingleCellMSC_SCD_Poster.pdf]</a></p>
<h3 id="visualization-r">Visualization-R</h3> 
<p><a href="https://www.ncbi.nlm.nih.gov/pubmed/29152409">Nature Biomedical Engineering paper - Designing probes for genome capture [URL: https://www.ncbi.nlm.nih.gov/pubmed/29152409]</a></p>
<h3 id="neural-networks-and-single-cell-transcriptomics">Neural networks</h3>
<p><a href="https://sivome.github.io/genomics/2019/11/17/Denoising-Single-Cell-Data-With-Autoencoders.html">Neural Networks blog - 10X single cell data [URL: https://sivome.github.io/genomics/2019/11/17/Denoising-Single-Cell-Data-With-Autoencoders.html]</a> 
<h3 id="jupyter">Jupyter</h3> 
<p><a href="https://sivome.github.io/proteomics/2019/03/11/Reporting_Proteomics_Results.html">Jupyter notebook - Summarizing proteomics results [URL: https://sivome.github.io/proteomics/2019/03/11/Reporting_Proteomics_Results.html]</a>
<h3 id="r-markdown">R Markdown</h3>
<p><a href="https://sivome.github.io/proteomics/2019/03/15/universe-of-proteomics-with-tidyverse.html">Tidyverse R - OMSSA [URL: https://sivome.github.io/proteomics/2019/03/15/universe-of-proteomics-with-tidyverse.html]</a>

<p>
<br />
</p>
