<!doctype html>
<html>
<head>
  <meta charset=utf-8 />
  <link href="media/vish-screen.css" type="text/css" rel="stylesheet" media="screen">
  <link href="media/vish-print.css" type="text/css" rel="stylesheet" media="print">
</head>
<body>
  <div id="main">
    <div id="content">
<h1 id="viswanadham-sridhara">Viswanadham (Vish) Sridhara</h1>
<h2 id="contact">CONTACT</h2>
<p>Viswanadham.Sridhara@gmail.com<br />
Tel: (724) 590 2170<br />
<p>Current role: Senior Bioinformatics Scientist<br />
<a href="https://scholar.google.com/citations?hl=en&amp;user=UBvhRHgAAAAJ&amp;view_op=list_works&amp;sortby=pubdate">Google Scholar</a> | <a href="https://www.linkedin.com/in/viswanadham-sridhara-308263174">LinkedIn</a> | <a href="https://github.com/viswam78">Github</a> | <a href="https://bitbucket.org/viswam78/">BitBucket</a> | <a href="https://sivome.github.io/">Blog</a> | <a href="https://hub.docker.com/u/sridharabio">DockerHub</a></p>
<h2 id="profile">PROFILE</h2>
<p>I am a computational scientist with modeling and simulation experience in engineering and life sciences domains. Post-PHD, I have 15+ years’ experience in building bioinformatics pipelines for both next-generation sequencing (NGS) and mass-spectrometry technologies. In my last role at Food and Drug Administration (FDA contractor), I built pipelines for single-cell transcriptomics work using cellranger and other R packages (Seurat, Monocle etc).</p><br />
<p>Recently, my focus is on using deep learning neural network and other machine learning models to analyze high-dimensional -omics datasets to infer associations in the data. I usually program in Python/R/MATLAB and typically use tools built for -omics domain for reproducible research. I’m well versed with R packages and python modules (for data science and machine learning).</p><br />
<p>Till date, my research work has resulted in ~20 peer-reviewed publications and numerous conference proceedings and poster presentations (google scholar citations: ~720, h-index=13).</p>
<h2 id="education">EDUCATION</h2>
<h3 id="section">2003 - 2007</h3>
<p>PhD, Electrical and Computer Engineering, ODU, Norfolk, VA<br />
<strong>Dissertation:</strong> Molecular dynamics simulations to probe effects of ultra-short high-intensity electric pulses on biological suspensions</p>
<h3 id="section-1">2001 - 2003</h3>
<p>MS, Electrical and Computer Engineering, ODU, Norfolk, VA<br />
<strong>Thesis:</strong> Monte carlo analysis of millimeter-wave GaN Gunn diodes</p>
<h2 id="technical-skills">TECHNICAL SKILLS</h2>
<h3 id="machine-learning">Machine learning</h3>
<p>Regression, Regularization, Hierarchical Clustering, Discriminant analyses<br />
<h3 id="programming">Programming</h3>
<p>Python (including data science and ML modules such as numpy, pandas, scikit-learn, matplotlib, keras), R (tidyverse, ML packages), MATLAB, Perl, C, C++, Fortran<br />
<h3 id="other-skills">Other skills</h3>
<p>Git, Snakemake, Docker, UNIX, MacOS, LaTeX<br />
<h3 id="engineering">Engineering</h3>
<p>Monte-carlo models [MS Thesis focus], Molecular dynamics simulations for atomic modeling [PHD Disseration focus] 
<h3 id="ngs">NGS</h3>
<p>Bulk RNA-seq: HISAT2, featureCounts, GATK, BWA, Bowtie, Samtools, BEDtools<br />
<p>Single cell RNA-seq: CellRanger, Seurat, Monocle3, fastMNN, SAVER<br /> 
<h3 id="mass-spec">Mass-spec</h3> 
<p>Byonic, PEAKS, Proteome Discoverer, OMSSA, MODa<br />
<h3 id="systems-biology">Systems Biology</h3>
<p>BioCyc, MetaCyc, SAINT 
<h3 id="md-simulations">MD Simulations</h3>
<p>GROMACS, NAMD, VMD</p>
<h2 id="professional-experience">PROFESSIONAL EXPERIENCE</h2>
<h3 id="jun-19---present">Jun ’19 - Present</h3>
<p><em>Senior Bioinformatics Scientist</em><br />
<strong>Argentys </strong></p>
<p> Universities and medical centers (current contract) Team</p>
<p>• NGS analyses of targeted LINE-1 retrotransposons data <br />
• NGS analyses to understand vaccine stability </p>    
<strong>Argentys (FDA contractor)</strong></p>
<p>Contractor at Food and Drug Administration (Jun’ 19 - Aug’ 20) CBER HIVE Team</p>
<p><strong>Informatics Projects at FDA</strong></p>
<p>• Helped in building RNA-seq pipeline for automated bulk processing of large-scale genomics datasets (fastp, HISAT2, featureCounts, DESEQ2)<br />
• Built single-cell RNA-seq pipeline for analyzing 10X single cell datasets (CellRanger, SAVER, Seurat/Monocle, fastMNN and other R packages)</p>
<p><strong>Sample Research Projects at FDA</strong></p>
<p>• Identifying Molecular Features of Neural Stem Cells in Varied Differentiation Stages<br />
• Single cell RNA-seq analyses to investigate sub-populations of bone-marrow and adipose derived multipotential stromal cells (MSCs) from 3 donors</p>
<h3 id="apr-16---present">Apr ’16 - Present</h3>
<p><em>Independent Contractor</em><br />
<strong>Universities, Hospitals and Startups</strong></p>
<p>Independent Consultant / Contractor (part of 2016, 2017 and 2018) for MGH, WRI, biofuel startup, Booz Allen Hamilton and ProteinMetrics<br />
• At Booz Allen Hamilton, helped in automating preprocessing of gene expression data (microarray and RNA-seq) for placental datasets using raw data stored in NCBI’s GEO repository (PAT - Placental Atlas Tool).<br />
• At Booz Allen Hamilton, improved analytics of differential expression, gene ontology and pathway analyses of the above preprocessed placental datasets (&gt;20 RNA-seq and &gt;300 microarray datasets).<br />
• At ProteinMetrics, tested ProteinMetrics suite of tools.<br /> 
• At Mass General Hospital, improved primer design for a proof-of-principle assay to capture long genomic regions.<br />
• At biofuel startup, built NGS pipeline and developed new computational methods for downstream analyses.<br />
• At biofuel startup, whole genome sequence (WGS) analyses, including variant analyses, copy number variation and strain specific analyses for biofuel production.<br />
• At biofuel startup, set-up in-house Jbrowse genome browser and Sequenceserver (blast+ web server).<br />
• At WRI, helped analyze diverse mass-spectrometry datasets (global proteomics, phosphoproteomics) of breast cancer tissue samples.<br />
• At WRI (in collaboration with PNNL), integrated proteogenomics data to identify markers in laser capture microdissected breast tumor samples (Preliminary work published in AACR).</p>
<h3 id="dec-12---apr-16">Dec ’12 - Apr ’16</h3>
<p><strong>THE UNIVERSITY OF TEXAS SYSTEM (UT AUSTIN, UTSW MEDICAL CENTER DALLAS)</strong></p>
<p><em>Computational Biologist</em> Aug ’15 - Apr ’16<br />
<strong>UT Southwestern Medical Center – PROTEOMICS CORE FACILITY</strong></p>
<p>• Protein identification and quantitation using bioinformatics tools (e.g., OMSSA, MASCOT, Sequest, ProteomeDiscoverer, MaxQuant, SAINT).<br />
• Bioinformatics proteomics consultant work with UT Southwestern investigators.</p>
<p><em>Research Engineering/Scientist Associate V</em> Dec ’12 - Aug ’15<br />
<strong>The University of Texas at Austin – CENTER FOR COMPUTATIONAL BIOLOGY AND BIOINFORMATICS</strong></p>
<p>• Built machine learning models using LASSO regression to predict bacterial growth conditions from <em>in-silico</em> metabolic fluxes generated from metabolic network based flux balance analyses (FBA) methods.<br />
• Proteomics and metabolic flux data analyses of mass-spectrometry based datasets. <br />
  a. Proteomics data is analyzed using new unrestricted PTM search methods. <br />
  b. Metabolic flux data is analyzed using FiatFlux software.<br />
• Helped multiple labs with proteomics bioinformatics consulting and short-term contracts.<br />
• Taught computational mass-spectrometry in Big Data Summer School organized by CCBB (Center for Computational Biology and Bioinformatics every summer – 4 day course)</p>
<h3 id="jan-12---dec-12">Jan ’12 - Dec ’12</h3>
<p><em>Scientific Programmer</em><br />
<strong>Samuel Roberts Noble Foundation</strong></p>
<p>• Helped integrated GenoLogics LIMS system with mass-spec equipment • Built pathway databases using MetaCyc and PathwayTools</p>
<h3 id="jun-07---jan-12">Jun ’07 - Jan ’12</h3>
<p><em>Postdoctoral Fellow</em><br />
<strong>National Center for Biotechnology Information (NCBI)</strong></p>
<p>• Part of NCBI OMSSA (Open Mass Spectrometry Search Algorithm) project:<br />
a. Improved noise filtering routines/filters to accommodate high-resolution mass-spectrometry data.<br />
b. Optimized algorithm for ETD data analysis (ion-types, precursor peak filtering).<br />
c. User support by addition of requested enzymes, modifications, and other iontypes required for quantitative proteomics (iTRAQ, TMT 6-plex etc.).<br />
d. Improved sensitivity by adjusting for mass offsets, precursor charge assignment for low-res data. Presented related mass-spectrometry work in ASMS ’08, ’09, ’10 and ’11.</p>
<p>• Integration of sequence data &amp; high-throughput phosphoproteomics data:<br />
a. Developed a new method (proof-of-principle) for PTM annotation on genomes (NCBI RefSeq) by using high-throughput proteomics data (PhosphoSite, PhosphoELM) and domain definitions of protein families within NCBI Conserved Domain Database (CDD).<br />
b. Apart from the large-scale annotation of conserved PTM sites on genomes, this method could facilitate identification of novel evolutionarily conserved posttranslational modifications.</p>
<p>• Developed a prototype for new class of peptide identification search algorithm OMSSA-SIM.<br />
a. Motivation is to remove most of the limitations faced by present generation sequence library search algorithms, for example, exact precursor mass match, unknown PTM’s, mutations, different enzymatic cleavage etc.<br />
b. Preliminary work in ASMS 2011 (oral presentation in computational mass-spectrometry).</p>
<h3 id="summer-06">Summer ’06</h3>
<p><em>Student Volunteer</em><br />
<strong>Old Dominion University - Aerospace Engineering</strong></p>
<p>• Helped improve programming drives for a test bogie run on MAGLEV (train using magnetic levitation).</p>
<h3 id="summer-04">Summer ’04</h3>
<p><em>Intern</em><br />
<strong>Siemens Limited. Siemens Vdo Automotive</strong></p>
<p>• Finite element analyses (2D) to improve fuel injection solenoid models using Ansoft software.</p>
<h3 id="aug-03---may-07">Aug ’03 - May ’07</h3>
<p><strong>OLD DOMINION UNIVERSITY (JOSHI LAB, FRANK REIDY RESEARCH CENTER FOR BIOELECTRICS)</strong></p>
<p><em>Graduate Research Assistant</em> Aug ’03 - May ’07<br />
<strong>COMPUTATIONAL RESEARCH LAB - BIOELECTRICS</strong></p>
<p>• Compute intense Molecular Dynamics (MD) simulations to understand multiple nanopore formation in lipid bilayers using advanced high performance computing environment.<br />
  • Microscopic calculation of local lipid membrane permittivities and diffusion coefficients using molecular dynamics simulations.<br />
  • MD simulations to understand effects of sodium (Na) and calcium (Ca) ions in pore formation within the biological suspensions.</p>
<p><em>Research Assistant V</em> May ’04 - Dec ’06<br />
<strong>FRANK REIDY RESEARCH CENTER FOR BIOELECTRICS</strong></p>
<p>• Time Domain Dielectric Transient Spectroscopy (TDDS) measurements to monitor intrastructural changes in leukaemic cells after nanosecond pulsed electric field application.</p>
<h2 id="selected-talks">SELECTED TALKS</h2>
<ol type="1">
<li><p>Large-scale analysis of PTMs @ UT-Austin, Thermo Fisher Scientific, MD Anderson Cancer Center</p></li>
<li><p>Using subspectral interval matches to make novel identifications of ETD tandem mass spectra @ ASMS 2011</p></li>
<li><p>Predicting bacterial growth conditions from bacterial physiology @ Center for Systems and Synthetic Biology, UT-Austin</p></li>
<li><p>Open Mass Spectrometry Search Algorithm @ George Washington University</p></li>
</ol>
<h2 id="publications">PUBLICATIONS</h2>
<ol type="1">
<li><p>Tosi, L., Sridhara, V., Yang, Y., et. al., “Engineered DNA Capture Probes for Massively Multiplexed Cloning of Kilobase-Sized Genome Regions”, Nature Biomedical Engineering, 2017;1. pii: 0092. doi: 10.1038/s41551-017-0092. Epub 2017 Jun 26.</p></li>
<li><p>Caglar, M. U., et. al., “The E. coli molecular phenotype under different growth conditions”, Scientific Reports 2017 Apr 18;7:45303. doi: 10.1038/srep45303.</p></li>
<li><p>Brown, C. W., Sridhara, V., Boutz, D. R., et. al., “Large-scale analysis of post-translational modifications in E. coli under glucoselimiting conditions”, BMC Genomics 2017 Apr 17;18(1):301. doi: 10.1186/s12864-017-3676-8.</p></li>
<li><p>Kalantari, R., Hicks, J., Li, L., Gagnon, K., Sridhara, V., Lemoff, A., Mirzaei, H., and Corey, D. R. “Stable Association of RNAi Machinery is Conserved Between the Cytoplasm and Nucleus of Human Cells”, RNA. 2016 Jul;22(7):1085-98. doi: 10.1261/rna.056499.</p></li>
<li><p>Houser, J. R., et. al., Controlled measurement and comparative analysis of cellular components in E. coli reveals broad regulatory changes under long-term starvation. PLOS Computational Biology 2015 Aug 14;11(8):e1004400.</p></li>
<li><p>Sridhara, V., Meyer AG, Rai P, Barrick JE, Ravikumar P, Segre D, and Wilke C.O. “ Predicting growth conditions from internal metabolic fluxes in an in-silico model of E. coli.” (PLoS One. 2014 Dec 12;9(12):e114608.).</p></li>
<li><p>Sridhara, V., Joshi, R. P. “Evaluations of a mechanistic hypothesis for the influence of extracellular ions on electroporation due to high-intensity, nanosecond pulsing.” Biochimica Biophysica Acta - Biomembranes, Vol. 1838, no. 7, Jul 2014, pp. 1793-1800.</p></li>
<li><p>Sridhara, V., Joshi, R. P. “Numerical study of lipid translocation driven by nanoporation due to multiple high-intensity, ultrashort electrical pulses.” Biochimica Biophysica Acta - Biomembranes, Vol. 1838, no. 7, Mar 2014, pp. 902-909.</p></li>
<li><p>Sridhara, V., Dina L. Bai, An Chi, Jeffrey Shabanowitz, Donald F. Hunt, Stephen H. Bryant and Lewis Y. Geer “Increasing peptide identifications and decreasing search times for ETD spectra by pre-processing and calculation of parent precursor charge.” Proteome Science 2012, 10:8.</p></li>
<li><p>Sridhara, V., Aron Marchler-Bauer, Stephen H. Bryant and Lewis Y. Geer, “Automatic annotation of experimentally derived, evolutionarily conserved post-translational modifications onto multiple genomes.” Database (Oxford), 2011, bar019.</p></li>
<li><p>Joshi, R.P., Sridhara, V., Song, J., Schoenbach, K.H “Aspects of Lipid Membrane BioResponses to Subnanosecond, Ultrahigh Voltage Pulsing.” IEEE Trans. Dielectrics &amp; Electrical Insulation, Vol. 16, no. 5, Oct 2009, pp. 1243-1250.</p></li>
<li><p>Allen L. Garner, George Chen, Nianyong Chen, Viswanadham Sridhara, Juergen F. Kolb, R. James Swanson, Stephen J. Beebe, Ravindra P. Joshi, and Karl H. Schoenbach “Changes in the Dielectric Properties of Jurkat Cells Induced by Ultrashort Electrical Pulses,” Biochemical Biophysical Research Communications Vol. 362, Issue. 1, Oct. 2007, pp. 139-144.</p></li>
<li><p>Joshi, R.P., Nguyen, A., Sridhara, V., Nuccitelli, R., and Schoenbach, K.H., “Simulations of Calcium Release Dynamics in Response to a High-Intensity, Ultra-Short Electric Pulse,” Physical review. E, Statistical, nonlinear, and soft matter physics 2007; 75(4 Pt 1):041920.</p></li>
<li><p>Pliquett, U., Joshi, R.P., Sridhara, V., Schoenbach, K.H. “High Electrical Field effects on Cell Membranes.” Bioelectrochemistry Vol. 70, Oct. 2006, pp. 275-282.</p></li>
<li><p>Joshi, R.P., Sridhara, V., Schoenbach, K.H “Microscopic Calculations of Local Lipid Membrane Permittivities and Diffusion Coefficients for Application to Electroporation Analyses.” Biochemical Biophysical Research Communications Vol. 348, Issue. 2, Sept. 2006, pp. 643648.</p></li>
<li><p>Qin Hu, Sridhara, V., R. P. Joshi, J. Kolb and K.H, Schoenbach, “Molecular Dynamics Analysis of High Electric Pulse effects on Bilayer Membranes Containing DPPC and DPPS.” (invited paper) IEEE Transactions on Plasma Science, Vol. 34, no. 4, Aug. 2006, pp. 1405-1411.</p></li>
<li><p>Hu, Q., Sridhara, V., Joshi, R.P., Schoenbach, K.H., Beebe, S.J., Blackmore, P.F. “Simulations of transient membrane behavior in cells subjected to a high-intensity ultrashort electric pulse.” Physical Review E - Statistical, Nonlinear, and Soft Matter Physics, Vol. 71,no. 3, March 2005, pp. 031914(1-7).</p></li>
<li><p>Joshi, R. P., Sridhara, V., Jogai, B., Shah, P., Del Rosario, R.D. “Analysis of dislocation scattering on electron mobility in GaN high electron mobility transistors.” Journal of Applied Physics, Vol. 93, no. 12, Jun 15, 2003, pp. 10046-10052.</p></li>
<li><p>Joshi, R. P., Sridhara, V., Shah, P., Del Rosario, R.D. “Monte Carlo analysis of GaN-based Gunn oscillators for microwave power generation.” Journal of Applied Physics, Vol. 93, no. 8, Apr 15, 2003, pp. 4836-4842.</p></li>
</ol>
<h2 id="conference-proceedings">CONFERENCE PROCEEDINGS</h2>
<ol type="1">
<li><p>Sridhara, V., Joshi, R.P., Hu, Q., Nguyen, A., and Schoenbach, K.H. “Simulations of cell response to high intensity, ultrashort electrical pulses” (invited paper) Bio-Electromagnetic Society Conference, Cancun, Mexico, June 11-15, 2006.</p></li>
<li><p>Pliquett, U., et. al., “Monitoring intrastructural changes in Jurkat cells after nanosecond pulsed electric field application.” Electromed 2005 Conference, Portland, OR, May 15-18 , 2005.</p></li>
<li><p>Garner, A.L., et. al., “Changes in the Dielectric Properties of Jurkat and HL-60 cells after Electrical Pulses.” Electromed 2005 Conference, Portland, OR, May 15-18, 2005.</p></li>
</ol>
<h2 id="posters">POSTERS</h2>
<p>AACR 2017, Proceedings of the AACR 2016, EMBL Conference 2014, Synthetic Biology: Engineering, Evolution and Design (SEED), ASMS 2010, ASMS 2009, ASMS 2008</p>
<h2 id="sample-projects">SAMPLE PROJECTS</h2>
<h3 id="lda">LDA</h3>
<p><a href="https://www.ncbi.nlm.nih.gov/pubmed/22321509">Database (Oxford) paper summarizing ML + bioinformatics [URL: https://www.ncbi.nlm.nih.gov/pubmed/22321509]</a>
<h3 id="ml-regularization">ML-Regularization</h3>
<p><a href="https://www.ncbi.nlm.nih.gov/pubmed/25502413">PLoS One paper - Regularization and metabolic network models [URL: https://www.ncbi.nlm.nih.gov/pubmed/25502413]</a>
<h3 id="visualization-r">Visualization-R</h3> 
<p><a href="https://www.ncbi.nlm.nih.gov/pubmed/29152409">Nature Biomedical Engineering paper - Designing probes for genome capture [URL: https://www.ncbi.nlm.nih.gov/pubmed/29152409]</a></p>
<h3 id="neural-networks-and-single-cell-transcriptomics">Neural networks</h3>
<p><a href="https://sivome.github.io/genomics/2019/11/17/Denoising-Single-Cell-Data-With-Autoencoders.html">Neural Networks blog - 10X single cell data [URL: https://sivome.github.io/genomics/2019/11/17/Denoising-Single-Cell-Data-With-Autoencoders.html]</a> 
<h3 id="jupyter">Jupyter</h3> 
<p><a href="https://sivome.github.io/proteomics/2019/03/11/Reporting_Proteomics_Results.html">Jupyter notebook - Summarizing proteomics results [URL: https://sivome.github.io/proteomics/2019/03/11/Reporting_Proteomics_Results.html]</a>
<h3 id="r-markdown">R Markdown</h3>
<p><a href="https://sivome.github.io/proteomics/2019/03/15/universe-of-proteomics-with-tidyverse.html">Tidyverse R - OMSSA [URL: https://sivome.github.io/proteomics/2019/03/15/universe-of-proteomics-with-tidyverse.html]</a>

<p>
<br />
</p>
