<!DOCTYPE html>
<html lang="en"><head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1"><!-- Begin Jekyll SEO tag v2.5.0 -->
<title>Single cell genomics - Seurat marker analyses | Sivome</title>
<meta name="generator" content="Jekyll v3.7.4" />
<meta property="og:title" content="Single cell genomics - Seurat marker analyses" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="Most of the RNA-seq experiments focus on bulk RNA-seq methods. However, after closely looking at single cell datasets, the information obtained from single-cell experiments can throw light on variety of underlying biological processes. Here, I downloaded publicly available microwell-seq dataset (Mouse Cell Atlas) that has 400K cells profiled. Out of these 400K cells, 242K cells seem to have meta data information. I picked top 10K cells for Seurat analyses in this blog." />
<meta property="og:description" content="Most of the RNA-seq experiments focus on bulk RNA-seq methods. However, after closely looking at single cell datasets, the information obtained from single-cell experiments can throw light on variety of underlying biological processes. Here, I downloaded publicly available microwell-seq dataset (Mouse Cell Atlas) that has 400K cells profiled. Out of these 400K cells, 242K cells seem to have meta data information. I picked top 10K cells for Seurat analyses in this blog." />
<link rel="canonical" href="http://localhost:4000/genomics/2019/09/05/Single-cell-genomics.html" />
<meta property="og:url" content="http://localhost:4000/genomics/2019/09/05/Single-cell-genomics.html" />
<meta property="og:site_name" content="Sivome" />
<meta property="og:type" content="article" />
<meta property="article:published_time" content="2019-09-05T00:00:00-04:00" />
<script type="application/ld+json">
{"headline":"Single cell genomics - Seurat marker analyses","dateModified":"2019-09-05T00:00:00-04:00","datePublished":"2019-09-05T00:00:00-04:00","url":"http://localhost:4000/genomics/2019/09/05/Single-cell-genomics.html","mainEntityOfPage":{"@type":"WebPage","@id":"http://localhost:4000/genomics/2019/09/05/Single-cell-genomics.html"},"description":"Most of the RNA-seq experiments focus on bulk RNA-seq methods. However, after closely looking at single cell datasets, the information obtained from single-cell experiments can throw light on variety of underlying biological processes. Here, I downloaded publicly available microwell-seq dataset (Mouse Cell Atlas) that has 400K cells profiled. Out of these 400K cells, 242K cells seem to have meta data information. I picked top 10K cells for Seurat analyses in this blog.","@type":"BlogPosting","@context":"http://schema.org"}</script>
<!-- End Jekyll SEO tag -->
<link rel="stylesheet" href="/assets/main.css"><link type="application/atom+xml" rel="alternate" href="http://localhost:4000/feed.xml" title="Sivome" /></head>
<body><header class="site-header" role="banner">

  <div class="wrapper"><a class="site-title" rel="author" href="/">Sivome</a><nav class="site-nav">
        <input type="checkbox" id="nav-trigger" class="nav-trigger" />
        <label for="nav-trigger">
          <span class="menu-icon">
            <svg viewBox="0 0 18 15" width="18px" height="15px">
              <path d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.032C17.335,0,18,0.665,18,1.484L18,1.484z M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.032C17.335,6.031,18,6.696,18,7.516L18,7.516z M18,13.516C18,14.335,17.335,15,16.516,15H1.484 C0.665,15,0,14.335,0,13.516l0,0c0-0.82,0.665-1.483,1.484-1.483h15.032C17.335,12.031,18,12.695,18,13.516L18,13.516z"/>
            </svg>
          </span>
        </label>

        <div class="trigger"><a class="page-link" href="/about/">About</a><a class="page-link" href="/cv/">CV</a></div>
      </nav></div>
</header>
<main class="page-content" aria-label="Content">
      <div class="wrapper">
        <article class="post h-entry" itemscope itemtype="http://schema.org/BlogPosting">

  <header class="post-header">
    <h1 class="post-title p-name" itemprop="name headline">Single cell genomics - Seurat marker analyses</h1>
    <p class="post-meta">
      <time class="dt-published" datetime="2019-09-05T00:00:00-04:00" itemprop="datePublished">Sep 5, 2019
      </time></p>
  </header>

  <div class="post-content e-content" itemprop="articleBody">
    <p>Most of the RNA-seq experiments focus on bulk RNA-seq methods. However, after closely looking at single cell datasets, the information obtained from single-cell experiments can throw light on variety of underlying biological processes. Here, I downloaded publicly available microwell-seq dataset (Mouse Cell Atlas) that has 400K cells profiled. Out of these 400K cells, 242K cells seem to have meta data information. I picked top 10K cells for Seurat analyses in this blog.</p>

<p>More R vignettes on Seurat can be found here: https://satijalab.org/seurat/vignettes.html
Mouse cell atlas (MCA): http://bis.zju.edu.cn/MCA/
Seurat R vignette on analyzing MCA can be found here: https://satijalab.org/seurat/v3.0/mca.html
Most of the code below is directly taken from the Seurat Vignette.</p>

<p>Seurat package can be used to classify different cell types and do additional analyses, such as finding markers that are specific to the cell types.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># required packages</span><span class="w">
</span><span class="n">rm</span><span class="p">(</span><span class="n">list</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ls</span><span class="p">())</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>
<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">Seurat</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">markdown</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">tinytex</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>Download the MCA processed data i.e., RDS file directly from the link found in Seurat MCA guided clustering and here I stored them in mca.matrix and mca.metadata.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Quick look at the meta data</span><span class="w">
</span><span class="n">head</span><span class="p">(</span><span class="n">mca.metadata</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>##                               ClusterID  Tissue     Batch
## Bladder_1.AAAACGAAAACGGGGCGA  Bladder_1 Bladder Bladder_1
## Bladder_1.AAAACGAAGCGGCCGCTA  Bladder_5 Bladder Bladder_1
## Bladder_1.AAAACGAAGTACTAGCAT Bladder_16 Bladder Bladder_1
## Bladder_1.AAAACGACGTTGCTGTGT  Bladder_8 Bladder Bladder_1
## Bladder_1.AAAACGAGCGAGCGAGTA  Bladder_4 Bladder Bladder_1
## Bladder_1.AAAACGAGGGTCAGATGG  Bladder_7 Bladder Bladder_1
##                                    Cell.Barcode
## Bladder_1.AAAACGAAAACGGGGCGA AAAACGAAAACGGGGCGA
## Bladder_1.AAAACGAAGCGGCCGCTA AAAACGAAGCGGCCGCTA
## Bladder_1.AAAACGAAGTACTAGCAT AAAACGAAGTACTAGCAT
## Bladder_1.AAAACGACGTTGCTGTGT AAAACGACGTTGCTGTGT
## Bladder_1.AAAACGAGCGAGCGAGTA AAAACGAGCGAGCGAGTA
## Bladder_1.AAAACGAGGGTCAGATGG AAAACGAGGGTCAGATGG
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Seurat analyses (Most of the content in this block is directly copied from the Seurat Vignette)</span><span class="w">
</span><span class="nf">dim</span><span class="p">(</span><span class="n">mca.matrix</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>## [1]  39855 405191
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Extract first 10K cells</span><span class="w">
</span><span class="n">mca.matrix.10K</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mca.matrix</span><span class="p">[,</span><span class="m">1</span><span class="o">:</span><span class="m">10000</span><span class="p">]</span><span class="w">
</span><span class="n">mca.10K</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">CreateSeuratObject</span><span class="p">(</span><span class="n">counts</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">mca.matrix.10K</span><span class="p">,</span><span class="w"> </span><span class="n">meta.data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">mca.metadata</span><span class="p">,</span><span class="w"> </span><span class="n">project</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"MouseCellAtlas"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Normalize data</span><span class="w">
</span><span class="n">mca.10K</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">NormalizeData</span><span class="p">(</span><span class="n">mca.10K</span><span class="p">,</span><span class="w"> </span><span class="n">normalization.method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"LogNormalize"</span><span class="p">,</span><span class="w"> </span><span class="n">scale.factor</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10000</span><span class="p">)</span><span class="w">
</span><span class="c1"># Variable features in the data</span><span class="w">
</span><span class="n">mca.10K</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">FindVariableFeatures</span><span class="p">(</span><span class="n">mca.10K</span><span class="p">)</span><span class="w">

</span><span class="c1"># Mitochondrial expression is not used in the variable feature list for classification</span><span class="w">
</span><span class="c1"># Mitochondrial genes generally start with 'mt-', but sometimes they could be in different case as well</span><span class="w">
</span><span class="n">mca.10K</span><span class="p">[[</span><span class="s2">"percent.mt"</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">PercentageFeatureSet</span><span class="p">(</span><span class="n">mca.10K</span><span class="p">,</span><span class="w"> </span><span class="n">pattern</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"^mt-"</span><span class="p">)</span><span class="w">
</span><span class="n">mca.10K</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ScaleData</span><span class="p">(</span><span class="n">mca.10K</span><span class="p">,</span><span class="w"> </span><span class="n">vars.to.regress</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"percent.mt"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>## Regressing out percent.mt
</code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>## Centering and scaling data matrix
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># dimensionality reduction using PCA, followed by t-SNE and UMAP analyses</span><span class="w">
</span><span class="c1"># This routine also involves finding neighbours and finding clusters</span><span class="w">
</span><span class="n">mca.10K</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">RunPCA</span><span class="p">(</span><span class="n">mca.10K</span><span class="p">,</span><span class="w"> </span><span class="n">npcs</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">100</span><span class="p">,</span><span class="w"> </span><span class="n">ndims.print</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">,</span><span class="w"> </span><span class="n">nfeatures.print</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>## PC_ 1
## Positive:  Gsta4, Wfdc2, Sptssb, Krt7, Sprr1a
## Negative:  Retnlg, Elane, Mpo, Srgn, Igkc
## PC_ 2
## Positive:  Retnlg, Elane, Mpo, Srgn, Igkc
## Negative:  Col3a1, Col1a2, Sparc, Dcn, Col1a1
## PC_ 3
## Positive:  Tagln, Myl9, Acta2, Myh11, Pcp4l1
## Negative:  Clec3b, Dpt, Htra3, Lum, Col14a1
## PC_ 4
## Positive:  Emcn, Plvap, Egfl7, Pecam1, Tmem88
## Negative:  Acta2, Myl9, Tagln, Myh11, Tpm2
## PC_ 5
## Positive:  Plvap, Emcn, Esam, Jam2, Ecscr
## Negative:  Tnc, Car3, Cxcl14, Rbp4, Dkk2
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">mca.10K</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">FindNeighbors</span><span class="p">(</span><span class="n">mca.10K</span><span class="p">,</span><span class="w"> </span><span class="n">reduction</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"pca"</span><span class="p">,</span><span class="w"> </span><span class="n">dims</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">75</span><span class="p">,</span><span class="w"> </span><span class="n">nn.eps</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">mca.10K</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">FindClusters</span><span class="p">(</span><span class="n">mca.10K</span><span class="p">,</span><span class="w"> </span><span class="n">resolution</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="n">n.start</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">#t-SNE, a populare classification method</span><span class="w">
</span><span class="n">mca.10K</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">RunTSNE</span><span class="p">(</span><span class="n">mca.10K</span><span class="p">,</span><span class="w"> </span><span class="n">dims</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">75</span><span class="p">)</span><span class="w">

</span><span class="c1">#UMAP is relatively new and with some datasets, it is shown to perform better than t-SNE</span><span class="w">
</span><span class="n">mca.10K</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">RunUMAP</span><span class="p">(</span><span class="n">mca.10K</span><span class="p">,</span><span class="w"> </span><span class="n">dims</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">75</span><span class="p">,</span><span class="w"> </span><span class="n">min.dist</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.75</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># PCA, TSNE, UMAP plots</span><span class="w">
</span><span class="c1"># Dark theme lets you see/count cells easily, and NoLegend ignores the labels of the clusters</span><span class="w">
</span><span class="n">PCAPlot</span><span class="p">(</span><span class="n">mca.10K</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">DarkTheme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">NoLegend</span><span class="p">()</span><span class="w">
</span></code></pre></div></div>
<p><img src="figure/PCA.png" alt="plot of chunk Dimplots" /></p>
<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">TSNEPlot</span><span class="p">(</span><span class="n">mca.10K</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">DarkTheme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">NoLegend</span><span class="p">()</span><span class="w">
</span></code></pre></div></div>
<p><img src="figure/TSNE.png" alt="plot of chunk Dimplots" /></p>
<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">UMAPPlot</span><span class="p">(</span><span class="n">mca.10K</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">DarkTheme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">NoLegend</span><span class="p">()</span><span class="w">
</span></code></pre></div></div>
<p><img src="figure/UMAP.png" alt="plot of chunk Dimplots" /></p>

<p>To visualize different tissues in the UMAP plot</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">DimPlot</span><span class="p">(</span><span class="n">mca.10K</span><span class="p">,</span><span class="w"> </span><span class="n">group.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'orig.ident'</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">DarkTheme</span><span class="p">()</span><span class="w">
</span></code></pre></div></div>

<p><img src="figure/Dimplot.png" alt="plot of chunk Dimplots" /></p>

<p>From the Dimplot, it is visually clear how different tissues get separated with UMAP classification</p>

<p>A general question research would be interested is the differences in the “Bladder” and “BoneMarrow”. The 10K cells I selected seems to included these 2 tissue types. There are other tissues in the original MCA dataset.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># To identify the markers between the bladder and the bone marrow</span><span class="w">
</span><span class="n">Markers_Bladder_BoneMarrow</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">FindMarkers</span><span class="p">(</span><span class="n">mca.10K</span><span class="p">,</span><span class="w"> </span><span class="n">group.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"orig.ident"</span><span class="p">,</span><span class="w"> </span><span class="n">ident.1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Bladder"</span><span class="p">,</span><span class="w"> </span><span class="n">ident.2</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"BoneMarrow"</span><span class="p">)</span><span class="w">
</span><span class="n">head</span><span class="p">(</span><span class="n">Markers_Bladder_BoneMarrow</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>##        p_val avg_logFC pct.1 pct.2 p_val_adj
## Mgp        0  4.880449 0.928 0.000         0
## Dcn        0  4.532426 0.763 0.000         0
## Sparc      0  4.124855 0.791 0.000         0
## Col1a2     0  4.037171 0.680 0.001         0
## Col3a1     0  3.919111 0.645 0.000         0
## Igfbp7     0  3.593033 0.862 0.003         0
</code></pre></div></div>

<p>This information is directly taken from the help file of the Seurat FindMarkers function:</p>

<p>avg_logFC: log fold-chage of the average expression between the two groups. Positive values indicate that the gene is more highly expressed in the first group</p>

<p>pct.1: The percentage of cells where the gene is detected in the first group</p>

<p>pct.2: The percentage of cells where the gene is detected in the second group</p>

<p>p_val_adj: Adjusted p-value, based on bonferroni correction using all genes in the dataset</p>

<p>Significant markers for bladder (i.e., ident.1) will be the genes that have a higher positive avg_logFC.
Similarly, markers for bonemarrow (i.e., ident.2) are the genes that have a higher negative avg_logFC</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Distribution of the avg_logFC (ideally volcano plot should be drawn with P-values as well)</span><span class="w">
</span><span class="n">hist</span><span class="p">(</span><span class="n">Markers_Bladder_BoneMarrow</span><span class="o">$</span><span class="n">avg_logFC</span><span class="p">,</span><span class="w"> </span><span class="m">25</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p><img src="figure/Hist.png" alt="plot of chunk unnamed-chunk-6" /></p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># For now, I used only the avg_LogFC values since there are significantly up- and down- regulated genes.</span><span class="w">
</span><span class="c1"># However, it is better to add p-value threshold as well!</span><span class="w">
</span><span class="n">significant_markers</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">Markers_Bladder_BoneMarrow</span><span class="p">[</span><span class="nf">abs</span><span class="p">(</span><span class="n">Markers_Bladder_BoneMarrow</span><span class="o">$</span><span class="n">avg_logFC</span><span class="p">)</span><span class="w"> </span><span class="o">&gt;</span><span class="m">4</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w">
</span><span class="n">significant_markers</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>##        p_val avg_logFC pct.1 pct.2 p_val_adj
## Mgp        0  4.880449 0.928 0.000         0
## Dcn        0  4.532426 0.763 0.000         0
## Sparc      0  4.124855 0.791 0.000         0
## Col1a2     0  4.037171 0.680 0.001         0
## Ltf        0 -4.181325 0.000 0.831         0
## Chil3      0 -4.270815 0.001 0.865         0
## Retnlg     0 -4.483080 0.000 0.891         0
## Lcn2       0 -4.890691 0.003 0.958         0
## Camp       0 -5.502772 0.000 0.938         0
## Ngp        0 -5.963624 0.000 0.972         0
## S100a9     0 -6.528570 0.001 0.998         0
## S100a8     0 -6.918372 0.001 1.000         0
</code></pre></div></div>
<p>To differentiate bladder and bone-marrow cells, Mgp, Dcn, Sparc and others with positive avg_logFC seem to be the markers for Bladder, while Ltf, Chil3, Retnlg and other genes with negative avg_logFC seem to be the markers for BoneMarrow.</p>

<p>The metadata will also let you compare using the group.by variable</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Comparison between Batches</span><span class="w">
</span><span class="n">Markers_BoneMarrow_1_and_4_batches</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">FindMarkers</span><span class="p">(</span><span class="n">mca.10K</span><span class="p">,</span><span class="w"> </span><span class="n">group.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Batch"</span><span class="p">,</span><span class="w"> </span><span class="n">ident.1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"BoneMarrow_1"</span><span class="p">,</span><span class="w"> </span><span class="n">ident.2</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"BoneMarrow_4"</span><span class="p">)</span><span class="w">
</span><span class="n">head</span><span class="p">(</span><span class="n">Markers_BoneMarrow_1_and_4_batches</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>##                  p_val avg_logFC pct.1 pct.2     p_val_adj
## Ftl1-ps1  0.000000e+00  2.328657 0.669 0.062  0.000000e+00
## Rps27rt   0.000000e+00  2.024656 0.471 0.013  0.000000e+00
## Gm12854   0.000000e+00  1.980651 0.392 0.005  0.000000e+00
## Gm14303  1.072558e-293  2.121924 0.527 0.037 4.274681e-289
## Gm5905   2.333793e-286  1.825317 0.498 0.031 9.301333e-282
## Gm10123  9.352412e-274  1.858993 0.384 0.011 3.727404e-269
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">significant_markers_batches</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">Markers_BoneMarrow_1_and_4_batches</span><span class="p">[</span><span class="nf">abs</span><span class="p">(</span><span class="n">Markers_BoneMarrow_1_and_4_batches</span><span class="o">$</span><span class="n">avg_logFC</span><span class="p">)</span><span class="w"> </span><span class="o">&gt;</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w">
</span><span class="n">significant_markers_batches</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>##                  p_val avg_logFC pct.1 pct.2     p_val_adj
## Ftl1-ps1  0.000000e+00  2.328657 0.669 0.062  0.000000e+00
## Rps27rt   0.000000e+00  2.024656 0.471 0.013  0.000000e+00
## Gm14303  1.072558e-293  2.121924 0.527 0.037 4.274681e-289
## Igha      8.891865e-46 -2.339435 0.004 0.305  3.543853e-41
</code></pre></div></div>

<p>When you compare between 2 batches, it seems that the Igha is up-regulated in 2nd sample i.e., BoneMarrow_4 batch. Likewise you can make other comparisons to identify the differentially expressed genes (i.e., markers) between different clusters!</p>

<p>In my next blog, I will use SingleR, an R package to subtype different cells / clusters.</p>

  </div><a class="u-url" href="/genomics/2019/09/05/Single-cell-genomics.html" hidden></a>
</article>

      </div>
    </main><footer class="site-footer h-card">
  <data class="u-url" href="/"></data>

  <div class="wrapper">

    <h2 class="footer-heading">Sivome</h2>

    <div class="footer-col-wrapper">
      <div class="footer-col footer-col-1">
        <ul class="contact-list">
          <li class="p-name">Sivome</li></ul>
      </div>

      <div class="footer-col footer-col-2"><ul class="social-media-list"><li><a href="https://github.com/Viswam78"><svg class="svg-icon"><use xlink:href="/assets/minima-social-icons.svg#github"></use></svg> <span class="username">Viswam78</span></a></li></ul>
</div>

      <div class="footer-col footer-col-3">
        <p>All -OMICS is eventually Sivomics!!</p>
      </div>
    </div>

  </div>

</footer>
</body>

</html>
